# ***Visualizing Operational Efficiency in Animal Shelters***

### **Milestone - 2 Clean, Manipulate, and Structure the data to build clear and informative visuals.**

In [1]:
#Importing all the required libraries
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns
import warnings

##  **Animal Intakes Data**

In [2]:
#Loading the Animal Intakes Data
intakes_data = pd.read_csv("Austin_Animal_Center_Intakes_20241105.CSV")

In [3]:
#Looking as to how the data looks like
intakes_data.head()

Animal ID          Name                DateTime     MonthYear  \
0   A786884        *Brock  01/03/2019 04:19:00 PM  January 2019   
1   A706918         Belle  07/05/2015 12:59:00 PM     July 2015   
2   A724273       Runster  04/14/2016 06:43:00 PM    April 2016   
3   A665644           NaN  10/21/2013 07:59:00 AM  October 2013   
4   A857105  Johnny Ringo  05/12/2022 12:23:00 AM      May 2022   

                        Found Location    Intake Type Intake Condition  \
0  2501 Magin Meadow Dr in Austin (TX)          Stray           Normal   
1     9409 Bluegrass Dr in Austin (TX)          Stray           Normal   
2   2818 Palomino Trail in Austin (TX)          Stray           Normal   
3                          Austin (TX)          Stray             Sick   
4   4404 Sarasota Drive in Austin (TX)  Public Assist           Normal   

  Animal Type Sex upon Intake Age upon Intake                     Breed  \
0         Dog   Neutered Male         2 years                Beagle Mix   
1         Dog   Spayed Female         8 years  English Springer Spaniel   
2         Dog     Intact Male       11 months               Basenji Mix   
3         Cat   Intact Female         4 weeks    Domestic Shorthair Mix   
4         Cat   Neutered Male         2 years        Domestic Shorthair   

          Color  
0      Tricolor  
1   White/Liver  
2   Sable/White  
3        Calico  
4  Orange Tabby

In [4]:
#Breaking Month Year into Month and Year as two separate columns
with warnings.catch_warnings():
    warnings.simplefilter("ignore")
    intakes_data[['Month','Year']] = intakes_data['MonthYear'].str.split(' ',expand=True)

In [5]:
#Breaking Found Location into City and State as two separate columns
with warnings.catch_warnings():
    warnings.simplefilter("ignore")

    # Defining a function
    def split_location(location):
        parts = location.rsplit(' ', 2)
        
        # If there's only one space, we treat the first part as City and the second as State
        if len(parts) == 2:
            city, state = parts
        elif len(parts) > 2:
            # if location has more than one space then split at the second last space
            city, state = parts[-2:]
        else:
            # If the format is unexpected, we are returning NaN
            city, state = np.nan, np.nan
            
        return pd.Series([city, state])

# Apply the function to the 'Found Location' column
intakes_data[['City', 'State']] = intakes_data['Found Location'].apply(split_location)

In [6]:
#Viewing the descriptive stats for the Animal Intakes data
intakes_data.describe()

Animal ID    Name                DateTime  MonthYear Found Location  \
count     168620  120090                  168620     168620         168620   
unique    151546   29161                  116247        134          68388   
top      A721033    Luna  09/23/2016 12:00:00 PM  June 2015    Austin (TX)   
freq          33     727                      64       2189          30660   

       Intake Type Intake Condition Animal Type Sex upon Intake  \
count       168620           168620      168620          168618   
unique           6               20           5               5   
top          Stray           Normal         Dog     Intact Male   
freq        115707           143111       92056           57128   

       Age upon Intake                   Breed        Color   Month    Year  \
count           168619                  168620       168620  168620  168620   
unique              55                    2972          651      12      12   
top             1 year  Domestic Shorthair Mix  Black/White     May    2019   
freq             27502                   33629        17477   17609   19727   

          City   State  
count   168620  168620  
unique      28       2  
top     Austin    (TX)  
freq    138226  166584

**For the objectives of this project, the Name, DateTime, MonthYear, Found Location, Breed, Color and State columns can be dropped.**

**Why?**

1. Name, DateTime, Breed and Color columns will not be necessary, as our analysis does not require data at such a granular level.

2. Found Location and MonthYear are being dropped as we have transformed and created new columns using the existing ones.

3. State is also being dropped because the dataset is from the State of Texas and all the data points will have state as Texas (TX).

In [7]:
intakes_data.drop(['Name','DateTime','MonthYear','Found Location','Breed','Color','State'], axis=1, inplace=True)

In [8]:
intakes_data.columns.to_list()

['Animal ID',
 'Intake Type',
 'Intake Condition',
 'Animal Type',
 'Sex upon Intake',
 'Age upon Intake',
 'Month',
 'Year',
 'City']

**Here we can see that the selected columns have been dropped from the data frame.**

In [9]:
intakes_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 168620 entries, 0 to 168619
Data columns (total 9 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Animal ID         168620 non-null  object
 1   Intake Type       168620 non-null  object
 2   Intake Condition  168620 non-null  object
 3   Animal Type       168620 non-null  object
 4   Sex upon Intake   168618 non-null  object
 5   Age upon Intake   168619 non-null  object
 6   Month             168620 non-null  object
 7   Year              168620 non-null  object
 8   City              168620 non-null  object
dtypes: object(9)
memory usage: 11.6+ MB


**Without even having to use the isna() function we can clearly see that out of 168,620 rows and 8 columns only two of them have missing values and that too the count is not very significant, only a couple of rows are missing and imputting them will not benefit the analysis significantly. So we are going to drop those two rows of data from intakes_data.**

In [10]:
intakes_data_cleaned = intakes_data.dropna(axis=0,how='any')

In [11]:
intakes_data_cleaned.isnull().any()

Animal ID           False
Intake Type         False
Intake Condition    False
Animal Type         False
Sex upon Intake     False
Age upon Intake     False
Month               False
Year                False
City                False
dtype: bool

***We can clearly see that none of the columns have missing values as we have dropped them.***

In [12]:
intakes_data_cleaned.tail(n=10)

Animal ID      Intake Type Intake Condition Animal Type  \
168610   A902459  Owner Surrender           Normal         Cat   
168611   A915712            Stray           Normal         Dog   
168612   A917339            Stray           Normal         Cat   
168613   A917373            Stray         Neonatal         Cat   
168614   A917347            Stray           Normal         Cat   
168615   A917365            Stray           Normal         Dog   
168616   A917357            Stray           Normal         Dog   
168617   A902460  Owner Surrender           Normal         Cat   
168618   A916886            Stray           Normal         Cat   
168619   A914535  Owner Surrender           Normal         Cat   

       Sex upon Intake Age upon Intake     Month  Year    City  
168610   Intact Female          1 year     April  2024  Austin  
168611     Intact Male          1 year  November  2024  Austin  
168612     Intact Male          1 year  November  2024  Travis  
168613         Unknown         4 weeks  November  2024  Austin  
168614     Intact Male        4 months  November  2024  Austin  
168615   Intact Female          1 year  November  2024  Austin  
168616   Intact Female          1 year  November  2024  Austin  
168617   Intact Female          1 year     April  2024  Austin  
168618     Intact Male         2 years   October  2024  Travis  
168619   Intact Female         1 month   October  2024  Austin

**The final cleaned data for the animal intakes is "intakes_data_cleaned".**

Note: Animal ID is a key assigned to each intake or when there has been any update to that record, at present the exploration to join this field to the Animal ID of outcomes data is yet to be done. As this falls outide the scope of the project at present.

##  **Animal Outcomes Data**

In [13]:
outcomes_data = pd.read_csv("Austin_Animal_Center_Outcomes_20241105.CSV")

In [14]:
outcomes_data.head()

Animal ID       Name                DateTime MonthYear Date of Birth  \
0   A882831  *Hamilton  07/01/2023 06:12:00 PM  Jul 2023    03/25/2023   
1   A794011      Chunk  05/08/2019 06:20:00 PM  May 2019    05/02/2017   
2   A776359      Gizmo  07/18/2018 04:02:00 PM  Jul 2018    07/12/2017   
3   A821648        NaN  08/16/2020 11:38:00 AM  Aug 2020    08/16/2019   
4   A720371      Moose  02/13/2016 05:59:00 PM  Feb 2016    10/08/2015   

  Outcome Type Outcome Subtype Animal Type Sex upon Outcome Age upon Outcome  \
0     Adoption             NaN         Cat    Neutered Male         3 months   
1    Rto-Adopt             NaN         Cat    Neutered Male          2 years   
2     Adoption             NaN         Dog    Neutered Male           1 year   
3   Euthanasia             NaN       Other          Unknown           1 year   
4     Adoption             NaN         Dog    Neutered Male         4 months   

                                Breed              Color  
0              Domestic Shorthair Mix        Black/White  
1              Domestic Shorthair Mix  Brown Tabby/White  
2             Chihuahua Shorthair Mix        White/Brown  
3                             Raccoon               Gray  
4  Anatol Shepherd/Labrador Retriever               Buff

In [15]:
#Breaking Month Year into Month and Year as two separate columns
with warnings.catch_warnings():
    warnings.simplefilter("ignore")
    outcomes_data[['Month','Year']] = outcomes_data['MonthYear'].str.split(' ',expand=True)

In [16]:
#Viewing the descriptive stats for the Animal Outcomes data
outcomes_data.describe()

Animal ID    Name                DateTime MonthYear Date of Birth  \
count     168536  120154                  168536    168536        168536   
unique    151457   29100                  140645       134          8518   
top      A721033    Luna  04/18/2016 12:00:00 AM  Jun 2019    05/01/2016   
freq          33     729                      39      2244           121   

       Outcome Type Outcome Subtype Animal Type Sex upon Outcome  \
count        168492           77383      168536           168534   
unique           11              26           5                5   
top        Adoption         Partner         Dog    Neutered Male   
freq          81294           39523       91985            59042   

       Age upon Outcome                   Breed        Color   Month    Year  
count            168525                  168536       168536  168536  168536  
unique               55                    2974          654      12      12  
top              1 year  Domestic Shorthair Mix  Black/White     Jul    2019  
freq              27962                   33995        17484   16506   19775

**For the objectives of this project, the Name, DateTime, Date of Birth, Breed and Color columns can be dropped.**

**Why?**

1. Name, DateTime, Date of Birth, Breed and Color columns will not be necessary, as our analysis does not require data at such a granular level.

2. MonthYear is being dropped as we have transformed and created new columns using the existing ones.

In [17]:
outcomes_data.drop(['Name','DateTime','MonthYear','Date of Birth','Breed','Color'],axis=1, inplace=True)

In [18]:
outcomes_data.columns.to_list()

['Animal ID',
 'Outcome Type',
 'Outcome Subtype',
 'Animal Type',
 'Sex upon Outcome',
 'Age upon Outcome',
 'Month',
 'Year']

**Here we can see that the five selected columns have been dropped from the data frame.**

In [19]:
outcomes_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 168536 entries, 0 to 168535
Data columns (total 8 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Animal ID         168536 non-null  object
 1   Outcome Type      168492 non-null  object
 2   Outcome Subtype   77383 non-null   object
 3   Animal Type       168536 non-null  object
 4   Sex upon Outcome  168534 non-null  object
 5   Age upon Outcome  168525 non-null  object
 6   Month             168536 non-null  object
 7   Year              168536 non-null  object
dtypes: object(8)
memory usage: 10.3+ MB


**We can clearly see that Outcome Subtype column is having a lot of missing values nearly more than 50% of the data is missing values, let's take a look at what unique values does this field store.**

In [20]:
print(outcomes_data['Outcome Subtype'].unique())

[nan 'Partner' 'Foster' 'Snr' 'SCRP' 'Out State' 'Suffering' 'At Vet'
 'Underage' 'Rabies Risk' 'In Kennel' 'Offsite' 'Aggressive' 'Field'
 'Emergency' 'Enroute' 'In Foster' 'Behavior' 'Medical'
 'Court/Investigation' 'Possible Theft' 'Barn' 'Customer S' 'In State'
 'Emer' 'In Surgery' 'Prc']


**We can eliminate this column as well alongside other columns as it isn't needed for our project.**

In [21]:
outcomes_data.drop("Outcome Subtype",axis=1,inplace=True)

In [22]:
#Checking for null values
outcomes_data.isnull().sum()

Animal ID            0
Outcome Type        44
Animal Type          0
Sex upon Outcome     2
Age upon Outcome    11
Month                0
Year                 0
dtype: int64

**Although, we have dropped the Outcome Subtype column we can still see that there are a couple of columns which have missing values in them. Rather than trying to imputing them, it would be more efficient to carry forward the analysis if we drop them as the count is not very high and not alarming.**

In [23]:
outcomes_data_cleaned = outcomes_data.dropna(axis=0,how='any')

In [24]:
#Checking for null values after dropping the null records
outcomes_data_cleaned.isnull().any()

Animal ID           False
Outcome Type        False
Animal Type         False
Sex upon Outcome    False
Age upon Outcome    False
Month               False
Year                False
dtype: bool

In [25]:
outcomes_data_cleaned.tail(n=10)

Animal ID     Outcome Type Animal Type Sex upon Outcome  \
168525   A917368         Transfer         Cat          Unknown   
168526   A916090         Adoption         Cat    Spayed Female   
168527   A916986         Adoption         Dog    Spayed Female   
168528   A917267  Return to Owner         Dog    Spayed Female   
168529   A916886         Adoption         Cat    Neutered Male   
168530   A915686         Adoption         Dog    Neutered Male   
168531   A902459         Adoption         Cat    Spayed Female   
168532   A902460         Adoption         Cat    Spayed Female   
168533   A914535         Adoption         Cat    Spayed Female   
168535   A916088         Adoption         Dog      Intact Male   

       Age upon Outcome Month  Year  
168525           4 days   Nov  2024  
168526         2 months   Nov  2024  
168527         2 months   Nov  2024  
168528          2 years   Nov  2024  
168529          2 years   Nov  2024  
168530           1 year   Nov  2024  
168531           1 year   Apr  2024  
168532           1 year   Apr  2024  
168533         2 months   Nov  2024  
168535          2 years   Nov  2024

**The final cleaned data for the animal outcomes is "outcomes_data_cleaned".**

Note: As stated in the Animal Intakes section, Animal ID can be joined with Intakes data to track the animal's status over time, but this falls out of the scope of the project. Once all the questions are answered this part of the equation is worth exploring.